### ISSUE TITLE
iOS - Planning

### ISSUE THREAD

**Issue by westnordost** [streetcomplete/StreetComplete#5421]
This ticket is something like the master ticket to coordinate development on an iOS port of StreetComplete. It replaces #1892 which also included a lot of discussion and research / observation work.

### TLDR: Have a look at the [Project Board](https://github.com/orgs/streetcomplete/projects/1) for a list of tasks. Contributions are welcome!

# Approach

The code base of this app is written in 100% Kotlin, a modern programming language that is quite similar to Swift. Kotlin code can be transpiled to JavaScript and also to machine code, just like Swift. Furthermore, many dependencies of this app by now are pure Kotlin libraries.

<img align="left" width="64" src="https://github.com/streetcomplete/StreetComplete/assets/4661658/b1c25b4c-2f60-463e-a587-12793db95961">

An iOS version of this app hence can be done with [Kotlin Multiplatform](https://www.jetbrains.com/kotlin-multiplatform/). 

The UI code can be shared using [Compose Multiplatform](https://www.jetbrains.com/lp/compose-multiplatform/), which is currently in alpha (= usable, but API may change). It is a fork of [Jetpack Compose](https://developer.android.com/jetpack/compose) with largely the same API. Jetpack Compose is a reactive UI framework in which developers define the UI completely in code (similar idea as SwiftUI). We'll have to re-create the entire UI (incrementally) in Compose, though.

This will keep the amount of code that needs to be platform dependent (Android / iOS) to a minimum.

This approach has the big advantage over others in that further maintenance of this project will not increase significantly, as there will continue to be just one code base.

# Steps in a nutshell

- Separate all platform specific code from application logic. Replace Android/Java dependencies for application logic with Kotlin multiplatform dependencies

- Incrementally migrate the UI code to Compose Multiplatform:

  1. <img align="right" width="64" src="https://github.com/streetcomplete/StreetComplete/assets/4661658/6f8ff785-140f-4360-9087-e811c71f7219">Migrate the UI code that currently uses Android XML layouts to Android [Jetpack Compose](https://developer.android.com/jetpack/compose). This can be done incrementally when done bottom-up (see [migration strategy guide](https://developer.android.com/jetpack/compose/migrate/strategy))

  2. <img align="right" clear="both" width="48" break="both" src="https://github.com/streetcomplete/StreetComplete/assets/4661658/6fe5a62e-e351-4366-a8b1-8ab6ca4c83f3"> Then, migrate from Jetpack Compose to Compose Multiplatform. With a little luck, Compose Multiplatform is out of alpha by then and thus the API more or less stable while the ecosystem around it is more mature.

# Current State

_This section will be updated continuously._

The steps necessary are described in detail in a Kanban-style [Project Board](https://github.com/orgs/streetcomplete/projects/1). Unless stated otherwise, they are explicitly up for the taking by people who read this and (I hope 😊) want to contribute to the development towards an iOS port. If you would like to implement one of the tasks described, comment in that ticket.

**Currently** the project board only contains tickets for stuff involving application logic, i.e. the point 1 from the "Steps in a nutshell" description from above. This is because I didn't look look deeper into with Jetpack Compose (Multiplatform) and their best practices yet, so decisions on which frameworks etc. to settle on have not been taken, yet. 

We very roughly estimated some time ago that developing a full port to iOS would take one man-year of work. So, the scope of this remains very large. We will only realistically achieve this together with lots of contributions from the community. Luckily, this is exactly what has been happening in the last years. Let's shift gears!

## How you can help

- You could take over a task from the project board! Or, you could familiarize yourself with the mentioned technologies / frameworks to make meaningful contributions later

- You can sponsor development! See the title "Sponsor this project" in the sidebar on the main page. More money allows me to devote more time to development on this app in general. If you know any eligible sources for larger scale funding, talk to me!

- You can help with maintenance and issue triage in general. Time saved on this is time I can put into furthering development on this!

**Comment by westnordost** [comment]
On what needs to be done on the UI side of things:

#### TL;DR: 
- Step 1 is #5070 
- Step 2 is to convert bottom up the view to Composables (Jetpack Compose)
- we'll see about the rest later

(Disclaimer: I (too) need to find my feet with those Jetpack Frameworks I haven't worked with yet, namely mostly Composables and Navigation. The below is a result of some research, looking at example code etc., so if something I did misunderstand something I describe below, please call it out!)

# Current UI architecture

![current_arch](https://github.com/streetcomplete/StreetComplete/assets/4661658/7801838d-7dff-4846-ba64-f887576f0c08)

Android *Activity*s are mostly just containers and *Fragment*s make up the brunt of components that define the view behavior, state, manage lifecycle, handle data (updates) and communication and navigation with/to their parent and child components.

Views are generally defined with Android XML layouts, a few "*ViewController" classes exist that define the behavior of certain view constructs in a reusable manner. Some compound/custom views may also (still) exist.

# Projected UI architecture

On Android, it would look something like this:

![target_arch](https://github.com/streetcomplete/StreetComplete/assets/4661658/7338e458-f41b-49d1-af2a-91a62344b6b8)

View layout, their behavior and their (ui) state is done via *Composable*s from Jetpack Compose / Compose Multiplatform.

The data used for display in the UI is accessed through *ViewModel*s. It is probably better to use Kotlin's `StateFlow` for the fields in the ViewModels instead of `LiveData` from the very start, because the former are already multiplatform.

Finally, the navigation in-between the different composables is managed using *NavHost* of Jetpack Navigation somehow (didn't read much into that yet). This means that the Activity is just the entry-point of the app and there are actually no Fragments left, everything is Composables and ViewModels, which means everything can be shared. (Hooray)

# UI architecture steps

Now, we must avoid at all costs to open up too many construction sites at the same time. It is not feasible to go from our current architecture to the other one in one big step. Instead, we must split up the migration in as many self-contained steps that make sure the codebase stays consistent in itself:

## Step 1. ViewModel

Slim Fragments: The data access and view state management shall go into **ViewModel**s, owned by the Fragments. We have a ticket for that already: #5070. Even independently of iOS, this makes sense, as it properly encapsulates view logic from accessing data / managing state.

## Step 2. Jetpack Compose
Convert the UI layout definition and behavior code bottom-up ([see migration guide](https://developer.android.com/jetpack/compose/migrate/strategy), [migration tutorial](https://developer.android.com/codelabs/jetpack-compose-migration#0)) to **Jetpack Compose**. This involves not only converting the Android XML layouts to Compose code, but also the code of "*ViewController"s, RecyclerView view holders and finally of course also Fragments, as that's where most of the view behavior is currently defined.

---

![current_view_arch drawio](https://github.com/streetcomplete/StreetComplete/assets/4661658/9094d77a-61ea-4330-a80a-9d8d9490a44b)
(After step 1 and 2)

## Step 3. Compose Multiplatform
Go the (hopefully small) step from Jetpack Compose to **Compose Multiplatform**. The biggest issue here will be how to access resources in a multiplatform way. There are some libraries already which facilitate this, like [MOKO resources](https://github.com/icerockdev/moko-resources), but hopefully a popular and stable solution will have emerged by then. I really hope there will be something official from JetBrains. (Alternative may also be a build-script that copies the resources over for iOS respectively.)

## Step 4. Multiplatform ViewModel / Navigation

Already by the end of step 2, Fragments will be very slim. We have the option to stop here view-architecture-wise and just duplicate the rest of the code that still lives in these Fragments on iOS in UIViewControllers defined in Swift. I think we should revisit this when we finished these steps, as these steps are really big on its own already anyway.

However, we still need to make at least ViewModels multiplatform. Interestingly, most libraries that provide some concept of ViewModels but as multiplatform also come with a solution to completely do away with Fragments (i.e. a replacement for *Navigation* and related things).

Those that provide a multiplatform replacement for only ViewModels are [KMM-ViewModel](https://github.com/rickclephas/KMM-ViewModel), [MOKO MVVM](https://github.com/icerockdev/moko-mvvm), [Summer](https://github.com/adevone/summer), [kmp-viewmodel](https://github.com/hoc081098/kmp-viewmodel) and others but most of these are either kind of alpha or older and quite inactive.

Here's a list of frameworks I had a little longer look at, all of these do something something ViewModel and also provide some glue for navigation:

### PreCompose

[PreCompose](https://github.com/Tlaster/PreCompose) provides both [*ViewModel*](https://github.com/Tlaster/PreCompose/blob/master/docs/component/view_model.md)s and [*Navigation*](https://github.com/Tlaster/PreCompose/blob/master/docs/component/navigation.md) whose API is pretty similar to the Android equivalent in Android. The project's readme itself mentions...

> If you familiar with Jetpack Lifecycle, ViewModel and Navigation, there will be nothing to learn.

Additionally, it has some [integration with *Koin*](https://github.com/Tlaster/PreCompose/blob/master/docs/component/koin.md), which is the dependency injection framework we use.

The project exists since 2021 and is continuously active since mid-2022, 0.6k people starred it on GitHub.

### Voyager

[Voyager](https://github.com/adrielcafe/voyager) also provides both for view models and navigation.

In particular, it introduces the concept of *Screen*s. A Screen conceptually is a replacement for a Activity or full-screen Fragment.

Instead of *ViewModel*s, we have [*ScreenModel*s](https://voyager.adriel.cafe/screenmodel), which have the same purpose as *ViewModel*s. Good thing: `ScreenModel` is just an interface. Also there is integration with Koin, the dependency injection framework we use.

Additionally, it has some setup for navigation with bottom sheets (which we use extensively) as well as tab navigation (used in profile screen).

The project exists since end of 2021 and continues to be very active, 1.8k people starred it on GitHub. Version 1.0.0 was released in December 2023 and shall bring API stability, the documentation is quite good. Of the three projects I've looked at, it is the youngest.

### Decompose

[Decompose](https://github.com/arkivanov/Decompose) can also serve as a replacement for the functionality described, although of the three libraries, it brings the greatest departure from the terminology, concepts and APIs from Android Jetpack:

- *Component*s are what I understand to be the replacement for the logic/navigation part of Fragments/Activities, as they optionally can have an associated *ComponentContext* (~ViewModel) if they need to be lifecycle-aware, keep their state during config changes and handle back button events
- *Value*s are observable fields within a component, pretty much what *LiveData or *StateFlow* would be
- *Child Stack* would be the replacement for the *FragmentManager*, i.e. something something navigation
- ....

One feature of Decompose is that it works well not only with *Compose* but also with *SwiftUI* and other UI frameworks, but I think we do not need that. But it means that Decompose is designed to be more of a general framework rather than being made for or tied to any specific UI framework.
Instead, it is written that *Decompose* goes well with *MVIKotlin*, another framework by the same author.

The project exists since 2020 (thus the oldest) and has been continuously very active (most active of those compared) ever since, 1.7k+ people starred it on GitHub. They are working on a new major version since at least December 2023. Striking is the extremely low number of open issues. 

### Conclusion

To be honest, I didn't really grasp the concepts of *Decompose* and the independency from Compose Multiplatform is nothing we would require. The migration to ViewModels + Composables + Navigation seems to me is already enough to chew, it does not help to then learn even new (even if related) concepts.

This is why I currently slightly favour Voyager, because just as PreCompose, it sticks quite closely to the concepts from Android Jetpack while at the same time is more popular and has more documentation + useful components than PreCompose has to offer, apparently.

The final decision on this lies somewhat in the future, though. The goal of this preliminary research has been to first ascertain whether the concepts of ViewModel + Navigation can be transferred to multiplatform too, so that the iterative steps 1 & 2 (before 3 & 4) are possible. The outcome: Yes, it's possible, frameworks exist whose concepts and APIs are quite similar to what we have with Android Jetpack.

**Comment by susrisha** [comment]
we on the other hand are also trying to make an iOS native version with my team https://github.com/TaskarCenterAtUW/GoInfoGame-iOS

**Comment by westnordost** [comment]
Interesting, first time I heard about "GoInfoGame". It looks like an abandoned fork of SC from 7 years ago that was created during a hackathon. So, your plan is to port that? Well.

Note that back then, SC was - architecturally, and what it could do - much more basic than now. Basically, one overpass query per quest type, display them on the map, answers are converted to OSM tags and uploaded. [I mentioned this old architecture in my talk on the SotM 2021 in the first few minutes](https://media.ccc.de/v/sotm2021-10044-how-streetcomplete-handles-edits#t=21), before I continued to talk about the current architecture. (Actually, that old architecture was already the next iteration from the overpass queries from 7 years ago, but nevermind)

People tend to wrongfully assume a project is simple because its UI is simple. So to your attempt, I would reply the same [as I did before in reply to someone else](https://github.com/streetcomplete/StreetComplete/issues/1892#issuecomment-1049856853) and add that GoMap!! (to my knowledge) already includes some basic StreetComplete-inspired quests and this is as far as one can go without, well, going towards replicating all of StreetComplete functionality eventually.
Of course, it goes without saying that trying to re-code something from scratch that has been continuously developed and refined over the course of 8+ years seems to be a futile endeavor, much less if there is already a plan well underway towards making it platform independent.

Anyway, if you made up your mind, does this mean that you do not anymore plan on helping with [porting the osmfeatures library to kotlin kmp](https://github.com/westnordost/osmfeatures/issues/22)?

(I will hide these comments as off topic as they are not related to planning an iOS port for this app)

**Comment by susrisha** [comment]
That was just an attempt we did and still pursuing for usage elsewhere. we still are pretty much inclined to helping with the port to kmp. Yeah the logic seems a bit old and we intended to do it as a separate endeavor.

**Comment by westnordost** [comment]
> Step 4. Multiplatform ViewModel / Navigation

Also,

- Android ViewModels are documented to be multiplatform starting with [v2.8.0](https://developer.android.com/jetpack/androidx/releases/lifecycle?s=09#2.8.0).

- JetBrains is [porting](https://www.jetbrains.com/help/kotlin-multiplatform-dev/compose-navigation-routing.html) AndroidX Navigation library to multiplatform. Currently experimental. 

To just keep using the AndroidX thingies would of course be the least effort.

**Comment by louwers** [comment]
Thanks for the great discussions at State of the Map EU. With regards to MapLibre Native for Compose Multiplatform, it seems someone already got it to work, someone else is building a library. 👍 

https://github.com/maplibre/maplibre-native/issues/2638

**Comment by Doomsdayrs** [comment]
Not exactly sure where to place, but this might be useful:

[Lifecycle path to Multiplatform](https://youtu.be/k1PIzEIO6jo)

**Comment by westnordost** [comment]
The funding through Prototype Fund ended, time for some progress update, retrospective and outlook!

# Progress Update

The work that has to be done can chiefly be grouped into two areas: 

1. **Replacing dependencies to Java**, Java-Libraries and Android with Kotlin Multiplatform (KMP) Libraries or own implementations on iOS
2. **Re-Implementing the whole UI in Compose**, also using ViewModels and Navigation.

### 1. Replacing dependencies to Java

The biggest tasks completed within the funding period in this regard were replacing the Java Http Client with the KMP library Ktor-Client, porting the [osmfeatures](https://github.com/westnordost/osmfeatures/) library to Kotlin Multiplatform and creating the [osm-opening-hours](https://github.com/westnordost/osm-opening-hours) Kotlin Multiplatform library. In detail:

- https://github.com/westnordost/osm-opening-hours
- #5483
- #5686
- #5656
- https://github.com/westnordost/osmfeatures/pull/25
- #5419
- using QRose and BigNum for QR Codes and parsing/generating them; writing an HTML parser to correctly display localized text styled with HTML, such as the changelog (both part of #5719)

![progress-dependencies](https://github.com/user-attachments/assets/1a445e08-6c3a-4580-8b2c-84482075c8e7)

(blue = already done before the funding period started)

What remains? Mostly fitzelkram and glue. But, well, that kind of stuff is often what takes longer than expected. One of the larger things  in terms of lines of code that still needs to be done are
- #5706 (in progress/on hold by @riQQ) and
- #5412 (I am holding off on this while other tasks in this area are still to do)

### 1. Re-Implementing the UI in Compose

...and introducing ViewModels + Navigation. Approximately, the absolute effort for this is about double that of the above, so this makes up the bulk of the to-be-done work.

#### MapLibre

An essential and large part of this undertaking was to migrate from the abandoned map renderer _tangram-es_ to _MapLibre-native_, done by @Helium314 and me. It was merged to master recently because we finally found a workaround for a critical error in MapLibre.

- #5693

#### Compose

I first focused on implementing all screens except the main screen. Recently, I broke forth into the main screen though, but mostly just the controls, dialogs etc.
In detail:

- #5719 
- #5607, #5855
- #5746
- #5799

![progress-ui](https://github.com/user-attachments/assets/54ed8cd3-06a6-4545-b131-88c48236bce8)

So, what's still to do are all the quest and overlay forms (and you know, they are often custom, that's why it is a lot of UI code) and abstracting the Android-specific implementation of MapLibre, i.e. use MapLibre in Compose. 

Now, I would love to not have to do the latter myself, but a library for this does not exist yet. There is, however, some developer interest in this. It looks like a group is forming to implement a library, but as it stands, there are currently 3-4 competing projects, so I hope there will be some consolidation soon: https://github.com/maplibre/maplibre-native/issues/2638

# Retrospective

The time estimations made for a full iOS-version of the app (one year full-time for one developer) are still quite on point.

The MapLibre integration took about 6 weeks more than expected even though @Helium314 did preliminary work on it before and then substantially helped me complete it. 
Re-Implementing the UI in Compose on the other hand turned out to be more straightforward and faster than expected, so that somewhat made up for the extra time spent on MapLibre.

Other than what has been mentioned, there were unfortunately no contributions made to this effort from other developers since I began to work full-time on it. So, now that I have worked on it for half a year, it would be roughly half a year more to complete this effort. (Plus, realistically, some generous buffer for unforeseen extra effort.)

# Outlook

### My to dos

Despite the funding having ended, there are a few things I still want to complete in order to get to a round conclusion of the work done so far. The time frame is 1-2 month (with less speed, need also some time out, really)

- stabilize v59
- upgrade to Jetpack Compose 1.7 and (try to) solve some things not yet implemented due to missing functionality in Jetpack Compose 1.6, such as reordering quest types in the settings and some animations in the statistics screen
- finish #5799 for v60. This one is a lot of effort, because it touches on so many components at once and includes all the stuff from the `MainActivity`, the erstwhile place where all the code for which I didn't find an abstraction, aka the garbage dump. It will also be a nightmare for @Helium314 to merge into SCEE, so that's why it is for v60 and not vor v59.
- get involved a bit - at least to make a few suggestions - with the group that wants to implement a MapLibre for Compose Multiplatform
- **maybe** #5412. I shy away from this right now because I fear it might open a new can of worms just when I wanted to come to a close

### Continuing

Extrapolating from the last half a year, I am now less hopeful about that it would be possible to complete this project by community contributions alone. The remaining work is also still substantial.
I will look into possible follow-up funding, but the outlook does not look too good: Another Prototype Fund will not be possible, last time I checked, OpenStreetMap has no money and the funding programme by the NLNet foundation is being dissolved in favor of some AI bullshit, too. Sponsors through Liberapay, Github sponsors and Patreon also didn't increase since this effort was announced, so crowd funding won't work either. Let's see. Suggestions welcome.

### Contributing

Of course, dear interested developer, any contribution done to this effort will just make completing it more feasible and no work towards that goal is lost, because we continuously integrate it into new versions of the app.

If you are inclined to help, the next steps in this process are to, bit by bit, reimplement the current UI in (Jetpack) Compose and use ViewModels. The migration to Compose happens in a bottom-up sort of fashion (widgets first, then parts of the screen, at the end, the whole screen). So, a most sustainable contribution would be to start with reimplementing widgets, such as the opening hours form (big effort), the hydrant diameter form (medium effort) or the building levels input form (small effort) (or many more).

Now, there could theoretically be a version of the app for iOS with limited functionality before we reached the 100%, by excluding those quests whose UI haven't been migrated yet. Tasks that have to definitely be done before that, however, are,

- MapLibre for Compose Multiplatform as a library (see https://github.com/maplibre/maplibre-native/issues/2638) and migration to that library

- #5412 and tasks that are currently blocked by this

---

As required by the Prototype Fund, I also created a report that will vanish in some archive that noone will read. Maybe you'd like to read it instead, however, it is in German and contains mostly the same stuff I wrote here: 
[Schlussbericht StreetComplete für iOS](https://github.com/user-attachments/files/16897777/01IS24S02.Schlussbericht.StreetComplete.fur.iOS.pdf)

**Comment by westnordost** [comment]
Oh, well, I didn't do that search yet because I first wanted to "clean up" by taking on the mentioned tasks, draw some breath and mind some other RL stuff that I didn't have the time to worry about in the last months. You know, enjoy the summer while it still lasts.
Also, I _do_ receive quite a bit of crowd funding IMO (for generally working on SC), so I really should do some quality of life improvements on the app again after I solely focused on this effort for so long.

There are a few organizations that seemed interested in pitching into sponsoring specifically an iOS version of the app, but IIRC each before I mentioned the scale of such an effort. Now that the scale has been reduced to about half, maybe I should ping them again, see what they think.
I have never asked OSM affiliated companies for financial support. I also can't imagine that they'd give out that kind of money like that, they'd rather give this directly to the foundation, I presume. I mean, they'd need it, too, OSM has very little money to spare.

Google would only allow for donations if it was done through their payment system. The app would then be marked as having "in-app payments", a very undesirable badge to have as a FOSS app. To aggressively ask Android users to donate for development of essentially an iOS version also sounds... not ideal. 😅

Applying for funding in my name would be a bad idea. But of course, know that you could apply in your name, too. Someone else with the necessary knowledge and skills could also implement it - I always encourage contributions. Anyway, I already got some potential funding sources bookmarked, feel free to email me with further suggestions.

**Comment by matkoniecz** [comment]
> Why should it make a difference whether I ask for support in your name or someone else's?

This kind of application for funding is basically "I promise to do XYZ work in exchange for receiving M amount of funding".

I would not want either for someone to make such promises on my behalf! 

Though if someone spotted potential well matching funding sources... Then I would be happy to learn about it! (Maybe in discussion section at this repository if it is extremely well fitting?)

**Comment by matkoniecz** [comment]
> Why is that undesirable?

Apps with such note are typically requiring payment to work, are blatant cash grabs or otherwise problematic. (With exception of ones where paying for something is whole point of app)

**Comment by matkoniecz** [comment]
> I have never asked OSM affiliated companies for financial support. I also can't imagine that they'd give out that kind of money like that, they'd rather give this directly to the foundation, I presume.

I think that at least in some cases there may be interest from them in funding very specific project. I would not assume that all such funding needs to be routed through OSMF.

**Comment by westnordost** [comment]
This is getting a bit off-topic, the ticket is about technical planning, not funding discussion, after all. It is not possible to transfer posts to a discussion on GitHub, so I'll just hide them here as off-topic. (We could continue in the discussions area.)

Charging for StreetComplete is off the table for me. StreetComplete is no convenience app that does anything _for_ the user, like e.g. a general purpose map app would. It may make it more convenient and fun to contribute to OSM, but contributing to OSM is still honorary engagement. It doesn't feel right to charge for the ability to do essentially unpaid work. Also, with projected usage numbers vs implementation effort, 1$ per user wouldn't cut it, by far.

@mcliquid, well, now I think you misunderstood me. I would be grateful for pointers to possible follow-up fundings. It happened plenty of times in the past that some people/orgs expressed interest in sponsoring and nothing came of it. I won't bet on that and I don't blame them when the extent is unclear - there is a huge difference between 1000 and 100000 Euros.
Anyway, what I meant with "you could apply": Others, like @matkoniecz , already applied for grants that had as its topic that they work on certain StreetComplete-stuff. I am fine with that.

**Comment by rugk** [comment]
First of all, thanks a lot for your efforts in any case!

> there were unfortunately no contributions made to this effort from other developers since I began to work full-time on it

One speculation I'd like to make about this: Most users in the community here (aka StreetComplete) likely are _Android_ users. So while I guess most of them see the point in a cross-platform app and supporting that case, fewer actually feel like contributing or would actively see the benefit (for themselves)?
And yeah, I know, no iOS device is needed in theory but yeah... people also want to see their efforts ideally. And for a kinds "low-level" project like this one, you just don't see the result (or only at the end, which is harder to reach).

As such, I agree having an MVP and a published iOS app would likely attackt iOS users and thus maybe also iOS developers or at least developers intended to supporting "their" "own" platform. After all, having an app to use by yourself is probably the best motivation for building it (in OSS world IMHO).

**Comment by rugk** [comment]
Another note on funding, and yes, I try to keep it short. Regarding:
>  the funding programme by the NLNet foundation is being dissolved in favor of some AI bullshit, too

Note if I understand it correctly, this is because the EU commission made the decicion to stop funding _Generation Internet initiative (NGI)_ resulting “in a loss of €27 million for software freedom.”

That said, **you can do something about it**, because they actually offer a public consultation, where you can criticize that decision!

More information and arguments here: https://fsfe.org/news/2024/news-20240911-02.html
(Deadline is: **20 September 2024**)

Only thing I want to add is: The target audience searched there that is likely most relevant here are "members of the public (e.g. pupils and students interested in an ICT-related career or employees in the digital field).". Especially, you can mention this app here or others and explain [it has been funded by that specific fund](https://nlnet.nl/project/StreetComplete/index.html), actually.

So make your voices heard!

**Comment by muescha** [comment]
> Other than what has been mentioned, there were unfortunately no contributions made to this effort from other developers since I began to work full-time on it.

I noticed there's no **CONTRIBUTING_IOS.md** file. When I want to contribute, I don’t know where to start, how to set up, or run the project—like which branch to use, what tools are needed, etc. It would be helpful if you could provide a step-by-step guide on setting up the project, or at least include links to tutorials to help get started.

**Comment by westnordost** [comment]
Currently, the project setup is still 100% an Android app as #5412 hasn't been done yet, so the prior project setup instructions apply. But yeah, in frame of solving the mentioned task, the documentation could/should be adapted and extended to include project setup for iOS.

**Comment by Finkregh** [comment]
Just a datapoint: ios user here and android in the past: the available apps on ios are not bad, but streetcomplete has been a much better experience ♥️

I can help as soon as testflight is available a d would be willing to sponsor (parts?) if the Apple developer thingy that costs money.

**Comment by Dinth** [comment]
> Just a datapoint: ios user here and android in the past: the available apps on ios are not bad, but streetcomplete has been a much better experience ♥️
> 
> I can help as soon as testflight is available a d would be willing to sponsor (parts?) if the Apple developer thingy that costs money.

Same here, i have migrated to iOS many years ago and while i was amazed how much better apps are (better design, more usability) on iOS, StreetComplete was the only app ive been missing ever since. Would love to chip in too.

**Comment by gr4v1ty69** [comment]
Chiming in to say I'm looking forward to this. I'm in no state a dev so can't really help, but would happily test if project completes. In the meantime, any alternative?

**Comment by mnalis** [comment]
> In the meantime, any alternative?

I'm not an iOS user, but [Go Map!!](https://wiki.openstreetmap.org/wiki/Go_Map!!) is usually mentioned in the context. And there is multiplatform [EveryDoor](https://wiki.openstreetmap.org/wiki/Every_Door) if you're looking for something that's easy to start up with. Or you could check a full list of [OSM Editors](https://wiki.openstreetmap.org/wiki/Editors) and take your pick.

**Comment by westnordost** [comment]
A short status update on what has been done since the last [status update](https://github.com/streetcomplete/StreetComplete/issues/5421#issuecomment-2332402123):

- in subsequent minor releases to **v59**, the migration to MapLibre ([#5693](https://github.com/streetcomplete/StreetComplete/pull/5693)) was stabilized 

- in **v60**, big parts of the main screen, including many dialogs, tutorial screens and menus, were migrated to Compose and subsequently stabilized ([#5799](https://github.com/streetcomplete/StreetComplete/pull/5799))

- I contributed to the incubating [maplibre-compose library](https://github.com/sargunv/maplibre-compose), a Compose Multiplatform wrapper for MapLibre, with the intent on migrating to it later.

- the UI of the building levels form was rebuilt in Compose by @GaeaKat ([#6022](https://github.com/streetcomplete/StreetComplete/pull/6022)). This is the first quest form that has been rebuilt in Compose! 🎉 It can serve as a good example for how the the migration should be done.

- @FloEdelmann recently started making database access multiplatform, WIP ([#6202](https://github.com/streetcomplete/StreetComplete/pull/6202))

So, It wasn't much, but is was honest work 🥔

**Comment by westnordost** [comment]
Good news! [NLNet is sponsoring development to fully migrate the app to Multiplatform!](https://nlnet.nl/project/StreetComplete-multiplatform/) Development on this from my side will soon accelerate again!

That said, the project plan / time estimations are still quite tight. So, as before, contributions are very much welcome - both monetary and helping out! Of course, 🖐️s are worth more than 💸, if you are able.

### 🧑‍💻 Do you want to lend a hand? 

- 📱Biggest part is the migration of the UI of all the quest and overlay forms to Compose. The PR #6022 serves as a fine example on how it should be done. (It can be split up in many small tasks)
  **Requires**: Some knowledge in Jetpack Compose

- 🗺️ Help stabilize [maplibre-compose](https://github.com/sargunv/maplibre-compose), especially for iOS. Check out their issue tracker. One feature specifically we need that is missing in maplibre-compose is https://github.com/sargunv/maplibre-compose/issues/22 .
  **Requires**: Good knowledge in Jetpack Compose, optimally a dev environment on a mac to test on iOS

- 🍏 iOS-only parts: NLNet was hesitant/ambivalent about sponsoring development specifically for iOS (rather than multiplatform in general), plus, I don't have much experience in it. Access to location, compass, accelerometer, app permissions; intra-app communication and a few other things might have to be implemented for iOS specifically (because there is no multiplatform library for that) - (but at the moment blocked by https://github.com/streetcomplete/StreetComplete/issues/5412 anyway)
  **Requires**: Knowledge and dev environment for iOS development


### 🦶 My next steps

I will probably do #5412 first and soon, in order to unblock many other tasks. Then, I will continuously work on reimplementing the UI of all the quests and overlays in Compose, step by step. (Which is, incidentally, also the largest group of tasks.)

**Comment by TailyFair** [comment]
Is plan to release once iOS port has all the features as android? If possible, I would love to try out asap, so even partial port would be great!

_...Return to bugging friends and family if they have spare android phone ^^..._

---
Edit:
I hope my comment did not seem too entitled, was just curious about release plan ^^
Got an Android phone, started contributing and it is amazing, great job!
Cannot help with code migration, will try to contribute by other means :)

**Comment by westnordost** [comment]
Yes, pretty much. (Sorry)

**Comment by theoricher** [comment]
Thank you so much for your work! I’m thrilled to see iOS compatibility making progress. The improvements this will bring to OSM, along with the influx of new contributors, are really exciting!

**Comment by pschorf** [comment]
Is there a good spot to jump in and contribute? It seems like a lot of the UI elements are already in flux in #6842, and I wasn't sure if the other parts of the project tracker (e.g. #6310) were too meaty to pick up. No worries if the help is more trouble than it's worth, just wanted to check in and see if you had any updated guidance for contributing.

**Comment by westnordost** [comment]
You are right about #6842 and pending UI work - there are no easy pickings anymore in that part. Whether certain tasks like #6310 would be too meaty to pick up or at what point you could help best depends on your experience, i.e. what languages and frameworks you have worked with so far. (And for #6310 in particular if there is maybe already a good Kotlin Multiplatform library available that would offer this functionality). In general, there are mostly bits and pieces not in place yet.

Assuming that you are an iOS developer (i.e. have a Mac), I think the best way to help, and also get a feel for the project, is to set up the dev environment on iOS and (try to) compile, see what comes up. Right now, the iOS app is just an empty shell but what you *could* do, is to set up a "mock" app where one could test already parts of the UI that have been migrated, such as the about screens and settings screens. I am sure tasks can be created from that, because surely not everything will work correctly out of the box there yet. (For example switching the language in-app is not supported in Compose, so on the Android side, changing the language requires some Android-specific code and also re-starting the ViewController)

**Comment by pschorf** [comment]
Awesome - sounds good. I'm very rusty but I do have access to a mac/iphone for testing, so I'll get started there! Thanks for all your hard work.

**Comment by allthetime** [comment]
Hi! I'm a iOS & React Native developer proficient in ObjC and Swift development - I do a lot of map based development with Mapbox & Maplibre and PostGIS backends. I'm 'good enough' with Android but not at the same level; I haven't done any work with Compose but this seems like a good place to jump in and learn. I contribute to OSM once in a while when I'm on bicycle trips and I love the idea of this project - gonna take it for a spin on my android when I'm out today. Thanks!!

**Comment by kiliankoe** [comment]
I've been watching this issue for a while and _yearn_ for an iOS port of StreetComplete, it's the one Android app I really miss. I have quite a bit of iOS-dev experience as well and would love to dig a bit deeper and figure out where/if I can contribute here :)

**Comment by westnordost** [comment]
So, basically what I wrote in my last comment: Other than migrating from MapLibre Android to [MapLibre Compose](https://github.com/maplibre/maplibre-compose) (#6072) which is somewhat reserved for @sargunv (or at least blocked by a certain missing feature in maplibre-compose), there are no big tasks left, but just many bits and pieces.

Which bits and pieces exactly will get somewhat clearer once we actually start compiling (pieces of) the app on iOS, running the tests etc.. Truth be told, I haven't done this yet as there are still enough other things to do in the migration and because my old MacBook 2013 just isn't up for the task anymore (need an Apple Silicon chip to compile) . I am only looking into a (second hand) replacement for it now.

**So, in a nutshell**: Set up the development environment on iOS, compile, run the unit tests. The iOS app is currently just an empty shell, you can bootstrap it with setting up koin (dependency injection) and a ViewController that basically just initializes Compose - from there, you could already test parts of the UI that have been fully migrated (such as the settings and about screens). Certainly, issues will pop up, for which new tickets/PRs can be created.

Also, what's always possible is to support me financially, of course, through Liberapay, Github Sponsors or Patreon.

**Comment by westnordost** [comment]
With #6842 **finally** merged, the remaining stuff has been somewhat cut down to stuff that can be counted on one's fingers.

Now, most of the UI should be testable on iOS from which I expect a swath of smaller iOS-specific issues to roll in. @kiliankoe already started testing and fixed some blocking issues. @kiliankoe, in case you are testing the quest UIs, maybe create a ticket in which to collect all the issues you notice in the UI, (including the small stuff, like "this looks unintuitive/alien for iOS users")? Other inclined iOS developers can then contribute issues they find into that ticket.

Apart from the elephant in the room (#6352), I previously mentioned it, there should be some more well-separable things that could be done to advance the migration.
I'll look around in the codebase and see if I can create some more tickets for these. Off the top of my head: location updates and requesting location permission (#6310), handle ongoing download/upload when app is sent to background.

**Comment by hejjoe** [comment]
Very glad to see the progress. If there is any chance to help with beta testing, feel free to post a link to the testflight app here. :)

**Comment by westnordost** [comment]
Alright, I looked around a bit and added two tickets, 

- #7031

- #7032

- #7081

which are tasks any developer that can test on iOS could pick up. The latter is quite small, the former might be a bit more effort.

**Comment by tim4tech** [comment]
Hi everyone! I just found this project and am excited to see if I can contribute. I wanted to implement something similar to this for my job but I think contributing to this project makes more sense. I have a lot of Kotlin, Swift, Compose, SwiftUI, mapping, OSM, etc skills, I see the Download and Upload issue is open above. Should I attempt to work on that?

**Comment by westnordost** [comment]
Hi Timothy, ha, that's exactly the right skill set! Indeed, few things remain that could be done by individual contributors. But the download & upload is still open. Note the comments on the ticket.

Another thing that will always be open, because I expect a long tail of issues pop up during testing, is to test the UI (for all the quests etc) on iOS. (All quest forms can be viewed in the settings screen. The settings screen would need to be added to the test scaffold for iOS) @kiliankoe started with that in #7000, but didn't find the time lately.

**Comment by westnordost** [comment]
**Check here for small (testing) tasks that people interested to help you take on**

# #7076

**Comment by westnordost** [comment]
This may sound odd because I have never tried to publish an iOS app before, but could it be that one of you that set up or tried to set up the app so far in their local development environment registered the bundle id `de.westnordost.streetcomplete` in their apple developer account? I am getting an unspecific error message when I try to add it to my account and I wonder whether the reason might be that someone else is "squatting" the id inadvertently (if that is even possible, I am not sure).

**Comment by muescha** [comment]
you have a detailed error message?

**Comment by sargunv** [comment]
I checked my account to verify my agent didn't do something like that, and it did not (though it does seem like bundle ids can be claimed without any sort of domain name verification? wild if that's squattable)

<img width="2506" height="1134" alt="Image" src="https://github.com/user-attachments/assets/f352d337-78e4-42c6-9321-e529f414b779" />

**Comment by kiliankoe** [comment]
It's just a string, by convention reverse-DNS is used, but it doesn't have to be. There's unfortunately no process here stopping anyone else from claiming one, but it's never shown in any user-facing UI either. And http://streetcomplete.westnordost.de/ is also not the project's landing page as it seems 😅

Xcode unfortunately just goes ahead and adds to that list if "Automatically manage signing" is turned on, so it could very much be that someone wanting to contribute to the project did that by accident. I'd suggest alternatively just going with `de.westnordost.streetcomplete.ios` or similar.

**Comment by westnordost** [comment]
Alright, no worries, I will just use `app.streetcomplete` then.

**Comment by sargunv** [comment]
https://github.com/streetcomplete/StreetComplete/pull/7088 is ready for review, and together with #6352, resolves #6072.

~~I can take #7081 and/or #7031 next?~~ now #7123 and #7124 are also ready for review.
